src/main/java/org/embulk/output/lib/GCSMediaHttpUploader.java

```diff
@@ -0,0 +1,939 @@
+package org.embulk.output.lib;
```

Comment:
MediaHttpUploader is created a new upload id every time the `upload` method is called. That doesn't allow us to resume the same upload-id. Let's say it try to upload from the beginning every time. 

https://github.com/vparfonov/google-api-java-client/blob/master/google-api-client/src/main/java/com/google/api/client/googleapis/media/MediaHttpUploader.java#L330

Reply:
Could you show me a diff from the original code?

Reply:
```
-    private HttpResponse resumableUpload(GenericUrl initiationRequestUrl) throws IOException {
+    private HttpResponse resumableUpload(GenericUrl initiationRequestUrl)
+            throws IOException
+    {
         // Make initial request to get the unique upload URL.
-        HttpResponse initialResponse = executeUploadInitiation(initiationRequestUrl);
-        if (!initialResponse.isSuccessStatusCode()) {
-            // If the initiation request is not successful return it immediately.
-            return initialResponse;
+        boolean isRetry = false;
+        if (uploadState == UploadState.NOT_STARTED || getUploadUrl() == null) {
+            HttpResponse initialResponse = executeUploadInitiation(initiationRequestUrl);
+            if (!initialResponse.isSuccessStatusCode()) {
+                // If the initiation request is not successful return it immediately.
+                return initialResponse;
+            }
+            try {
+                uploadUrl = new GenericUrl(initialResponse.getHeaders().getLocation());
+            }
+            finally {
+                initialResponse.disconnect();
+            }
         }
-        GenericUrl uploadUrl;
-        try {
-            uploadUrl = new GenericUrl(initialResponse.getHeaders().getLocation());
-        } finally {
-            initialResponse.disconnect();
+        else if (totalBytesClientSent > totalBytesServerReceived) {
+            // reset to the last sent file
+            contentInputStream.reset();
         }

         // Convert media content into a byte stream to upload in chunks.
-        contentInputStream = mediaContent.getInputStream();
+
         if (!contentInputStream.markSupported() && isMediaLengthKnown()) {
             // If we know the media content length then wrap the stream into a Buffered input stream to
             // support the {@link InputStream#mark} and {@link InputStream#reset} methods required for
@@ -398,15 +405,14 @@ public final class MediaHttpUploader {
         while (true) {
             currentRequest = requestFactory.buildPutRequest(uploadUrl, null);
             setContentAndHeadersOnCurrentRequest();
+            totalBytesClientSent = totalBytesClientSent + chunkSize;
             // set mediaErrorHandler as I/O exception handler and as unsuccessful response handler for
-            // calling to serverErrorCallback on an I/O exception or an abnormal HTTP response
-            new MediaUploadErrorHandler(this, currentRequest);
-
             if (isMediaLengthKnown()) {
                 // TODO(rmistry): Support gzipping content for the case where media content length is
                 // known (https://code.google.com/p/google-api-java-client/issues/detail?id=691).
                 response = executeCurrentRequestWithoutGZip(currentRequest);
-            } else {
+            }
+            else {
                 response = executeCurrentRequest(currentRequest);
             }
             boolean returningResponse = false;
@@ -416,7 +422,7 @@ public final class MediaHttpUploader {
                     if (mediaContent.getCloseInputStream()) {
                         contentInputStream.close();
                     }
-                    updateStateAndNotifyListener(com.google.api.client.googleapis.media.MediaHttpUploader.UploadState.MEDIA_COMPLETE);
+                    updateStateAndNotifyListener(UploadState.MEDIA_COMPLETE);
                     returningResponse = true;
                     return response;
                 }
@@ -451,7 +457,8 @@ public final class MediaHttpUploader {
                         long actualSkipValue = contentInputStream.skip(currentBytesServerReceived);
                         Preconditions.checkState(currentBytesServerReceived == actualSkipValue);
                     }
-                } else if (copyBytes == 0) {
+                }
+                else if (copyBytes == 0) {
                     // server got all the bytes, so we don't need to use this buffer. Otherwise, we have to
                     // keep the buffer and copy part (or all) of its bytes to the stream we are sending to the
                     // server
@@ -459,8 +466,9 @@ public final class MediaHttpUploader {
                 }
                 totalBytesServerReceived = newBytesServerReceived;

             if (isMediaLengthKnown()) {
+        boolean isRetry = false;
+        if (uploadState == UploadState.NOT_STARTED || getUploadUrl() == null) {
+            HttpResponse initialResponse = executeUploadInitiation(initiationRequestUrl);
+            if (!initialResponse.isSuccessStatusCode()) {
+                // If the initiation request is not successful return it immediately.
+                return initialResponse;
+            }
+            try {
+                uploadUrl = new GenericUrl(initialResponse.getHeaders().getLocation());
+            }
+            finally {
+                initialResponse.disconnect();
+            }
         }
-        GenericUrl uploadUrl;
-        try {
-            uploadUrl = new GenericUrl(initialResponse.getHeaders().getLocation());
-        } finally {
-            initialResponse.disconnect();
+        else if (totalBytesClientSent > totalBytesServerReceived) {
+            // reset to the last sent file
+            contentInputStream.reset();
         }

         // Convert media content into a byte stream to upload in chunks.
-        contentInputStream = mediaContent.getInputStream();
+
         if (!contentInputStream.markSupported() && isMediaLengthKnown()) {
             // If we know the media content length then wrap the stream into a Buffered input stream to
             // support the {@link InputStream#mark} and {@link InputStream#reset} methods required for
@@ -398,15 +405,14 @@ public final class MediaHttpUploader {
         while (true) {
             currentRequest = requestFactory.buildPutRequest(uploadUrl, null);
             setContentAndHeadersOnCurrentRequest();
+            totalBytesClientSent = totalBytesClientSent + chunkSize;
             // set mediaErrorHandler as I/O exception handler and as unsuccessful response handler for
-            // calling to serverErrorCallback on an I/O exception or an abnormal HTTP response
-            new MediaUploadErrorHandler(this, currentRequest);
-
             if (isMediaLengthKnown()) {
                 // TODO(rmistry): Support gzipping content for the case where media content length is
                 // known (https://code.google.com/p/google-api-java-client/issues/detail?id=691).
                 response = executeCurrentRequestWithoutGZip(currentRequest);
```

Reply:
```
  private void setContentAndHeadersOnCurrentRequest() throws IOException {
+    private void setContentAndHeadersOnCurrentRequest()
+            throws IOException
+    {
         int blockSize;
         if (isMediaLengthKnown()) {
             // We know exactly what the blockSize will be because we know the media content length.
             blockSize = (int) Math.min(chunkSize, getMediaContentLength() - totalBytesServerReceived);
-        } else {
+        }
+        else {
             // Use the chunkSize as the blockSize because we do know what what it is yet.
             blockSize = chunkSize;
         }

         AbstractInputStreamContent contentChunk;
         int actualBlockSize = blockSize;
-        if (isMediaLengthKnown()) {
-            // Mark the current position in case we need to retry the request.
-            contentInputStream.mark(blockSize);
-
-            InputStream limitInputStream = ByteStreams.limit(contentInputStream, blockSize);
-            contentChunk = new InputStreamContent(
-                    mediaContent.getType(), limitInputStream).setRetrySupported(true)
-                    .setLength(blockSize).setCloseInputStream(false);
-            mediaContentLengthStr = String.valueOf(getMediaContentLength());
-        } else {
-            // If the media content length is not known we implement a custom buffered input stream that
-            // enables us to detect the length of the media content when the last chunk is sent. We
-            // accomplish this by always trying to read an extra byte further than the end of the current
-            // chunk.
-            int actualBytesRead;
-            int bytesAllowedToRead;
-
-            // amount of bytes which need to be copied from last chunk buffer
-            int copyBytes = 0;
-            if (currentRequestContentBuffer == null) {
-                bytesAllowedToRead = cachedByte == null ? blockSize + 1 : blockSize;
-                currentRequestContentBuffer = new byte[blockSize + 1];
-                if (cachedByte != null) {
-                    currentRequestContentBuffer[0] = cachedByte;
-                }
-            } else {
-                // currentRequestContentBuffer is not null that means one of the following:
-                // 1. This is a request to recover from a server error (e.g. 503)
-                // or
-                // 2. The server received less bytes than the amount of bytes the client had sent. For
-                // example, the client sends bytes 100-199, but the server returns back status code 308,
-                // and its "Range" header is "bytes=0-150".
-                // In that case, the new request will be constructed from the previous request's byte buffer
-                // plus new bytes from the stream.
-                copyBytes = (int) (totalBytesClientSent - totalBytesServerReceived);
-                // shift copyBytes bytes to the beginning - those are the bytes which weren't received by
-                // the server in the last chunk.
-                System.arraycopy(currentRequestContentBuffer, currentChunkLength - copyBytes,
-                        currentRequestContentBuffer, 0, copyBytes);
-                if (cachedByte != null) {
-                    // add the last cached byte to the buffer
-                    currentRequestContentBuffer[copyBytes] = cachedByte;
-                }
-
-                bytesAllowedToRead = blockSize - copyBytes;
-            }
-
-            actualBytesRead = ByteStreams.read(
-                    contentInputStream, currentRequestContentBuffer, blockSize + 1 - bytesAllowedToRead,
-                    bytesAllowedToRead);
-
-            if (actualBytesRead < bytesAllowedToRead) {
-                actualBlockSize = copyBytes + Math.max(0, actualBytesRead);
-                if (cachedByte != null) {
-                    actualBlockSize++;
-                    cachedByte = null;
-                }

-                if (mediaContentLengthStr.equals("*")) {
-                    // At this point we know we reached the media content length because we either read less
-                    // than the specified chunk size or there is no more data left to be read.
-                    mediaContentLengthStr = String.valueOf(totalBytesServerReceived + actualBlockSize);
-                }
-            } else {
-                cachedByte = currentRequestContentBuffer[blockSize];
-            }
+        // Mark the current position in case we need to retry the request.
+        contentInputStream.mark(blockSize);

-            contentChunk = new ByteArrayContent(
-                    mediaContent.getType(), currentRequestContentBuffer, 0, actualBlockSize);
-            totalBytesClientSent = totalBytesServerReceived + actualBlockSize;
-        }
+        InputStream limitInputStream = ByteStreams.limit(contentInputStream, blockSize);
+        contentChunk = new InputStreamContent(
+                mediaContent.getType(), limitInputStream).setRetrySupported(true)
+                .setLength(blockSize).setCloseInputStream(false);
+        mediaContentLengthStr = String.valueOf(getMediaContentLength());

         currentChunkLength = actualBlockSize;
         currentRequest.setContent(contentChunk);
         if (actualBlockSize == 0) {
             // special case of zero content media being uploaded
             currentRequest.getHeaders().setContentRange("bytes */0");
```

Reply:
```
+    public GenericUrl getUploadUrl()
+    {
+        return uploadUrl;
+    }
```

Reply:
Could you post on gist with the below process? I can't catch up it enough because the diff is huge, isn't well-formatted and doesn't contain entire codes
1. Paste original entire codes of the class file
2. Save it
3. Paste diff that you changed
4. Save

e.g. https://gist.github.com/sakama/c9bdef090fc0191d018e66da9faf7c91/revisions

Reply:
Thank for your suggestion. That should be more clear.

https://gist.github.com/hieudion/5e943988e7864b519e6a53e206b275a9